sklearn/neighbors/classification.py

```diff
@@ -364,27 +383,51 @@ def predict(self, X):
 
         weights = _get_weights(neigh_dist, self.weights)
 
-        y_pred = np.empty((n_samples, n_outputs), dtype=classes_[0].dtype)
-        for k, classes_k in enumerate(classes_):
-            pred_labels = np.zeros(len(neigh_ind), dtype=object)
-            pred_labels[:] = [_y[ind, k] for ind in neigh_ind]
-            if weights is None:
-                mode = np.array([stats.mode(pl)[0]
-                                 for pl in pred_labels[inliers]], dtype=np.int)
-            else:
-                mode = np.array([weighted_mode(pl, w)[0]
-                                 for (pl, w)
-                                 in zip(pred_labels[inliers], weights[inliers])],
-                                dtype=np.int)
+        if self.outputs_2d_ == 'sparse':
+            y_pred_sparse_multilabel = []
 
-            mode = mode.ravel()
+            for k, classes_k in enumerate(classes_):
+                pred_labels = np.array([_y[ind, k].toarray()
```

Comment:
why this is converted to `object`. if we have a sparse matrix, we have for sure numerical values.
Then we don't even have to construct a dense matrix. we can just pass `*.data` from the sparse matrix

Reply:
In this case we need also to slice the weights

Reply:
So this is extracting a single column for all of the neighbors of each query. We are making it dense here, at a cost of (n_samples * avg_neighbors) memory. The sparse representation is to avoid a dense matrix of (n_samples * n_classes) where n_classes is presumed to be large. Yes, we can avoid making this dense, but I don't think it's especially problematic as long as avg_neighbours << n_classes.

We should be careful about constructing arrays of arrays/lists though. When the lengths of the arrays are all the same, this becomes a 2d array whose elements are ints, but still of dtype object due to this constructor. The safe way to do such construction is with np.empty(n, dtype='O') and then fill the array. But I don't see why pred_labels should be an array here at all.

Now that I've clarified that to myself, I'll go see what https://github.com/massich/scikit-learn/pull/5.

Reply:
Conclusion? we materialize the matrix here and later we can modify the mode computation propagating the sparse matrix later if needed. 
If so, we could use `mode = 1 if ( np.product(x.shape) - x.nnz ) < x.nnz else 0`

Reply:
> We should be careful about constructing arrays of arrays/lists though.

That is correct, so in the unlikely cornercase that each sample has the same number of neighbours, the current code is probably broken.

> But I don't see why pred_labels should be an array here at all.

The reason is because further down below, we use fancy numpy indexing with a list on this array, which does not work for a list (in `pred_labels[inliers]`). That can of course be converted to list comprehension, but might have performance bottleneck (not sure this is the case for object array)

Reply:
Actually, the ` zip(pred_labels[inliers], weights[inliers])]` works fine as well if it is a 2D array instead of 1D object array? So I think we can leave this as is.